### ISSUE TITLE
Prefer using PointerEvents

### ISSUE THREAD

**Issue by IvanSanchez** [Leaflet/Leaflet#7200]
TODO: Finish the [work on the `pointerify` branch](https://github.com/Leaflet/Leaflet/commit/3151aadb17a64131ecaa533a1f328c6a41339ff1), now that [the majority (~90%) of browsers support `PointerEvent`s](https://caniuse.com/#search=pointerevent):

- [ ] Make `DomEvent.on`/`off` display a deprecation warning when listening to a `MouseEvent` or `TouchEvent`.
- [ ] Rebase the [code](https://github.com/Leaflet/Leaflet/commit/3151aadb17a64131ecaa533a1f328c6a41339ff1) that polyfills the `DomEvent` stuff.
- [ ] Change all internal references to `MouseEvent`/`TouchEvent`s into `PointerEvent`s (e.g. turn `this._container.on('mousedown', ...` into `this._container.on('pointerdown', ...`)
- [ ] Ensure documentation reflects the changes (e.g. remove https://leafletjs.com/reference-1.6.0.html#mouseevent *et al*)

*Hopefully* this should remove/simplify the legacy logic for touch events. This is not tied to a particular issue, but things like #7027, #7199, #7097 should benefit from it

**Comment by johnd0e** [comment]
The idea is tempting, but I have some concerns.

1. Besides Leaflet itself we have **a lot of** 'legacy' plugins code.
Are you proposing to continue support Touch- and Mouse- by implementing internal polyfills?
If so, I'm not sure that this is right way, as it would be equally error-prone as it is now.
Instead, Leaflet could left this 'untouched': use Pointer- internally, but let all 3rd-party code to prefer any they want. 

2. There are some cases where Pointer-api does not correspond to Touch-api directly.
E.g. we can easily prevent 'click' or 'contextmenu' in 'touchend' listener, but it's not possible in 'pointerup' listener.

It's just thoughts, as I do not have ultimate solution for all current issues.
But you may look to my own approach in this [project](https://github.com/Leaflet/Leaflet/projects/1), where I try to simplify the things step by step.

**Comment by IvanSanchez** [comment]
> Are you proposing to continue support Touch- and Mouse- by implementing internal polyfills?

Yes. See https://github.com/Leaflet/Leaflet/commit/3151aadb17a64131ecaa533a1f328c6a41339ff1#diff-1917890cbccb15af367e6f7de954f297R147 - i.e. when an event handler expects a `TouchEvent`, the `DomEvent` code sends that handler a `TouchEvent`.

>  If so, I'm not sure that this is right way, as it would be equally error-prone as it is now.

I disagree here. The goal in my head is to have the straightforward code path (send a `PointerEvent` fired natively by the browser directly into a handler function that expects a `PointerEvent`) working as widely as possible. Then we can worry about Safari 12 and Firefox+Android and IE10.

I know this issue is brain-melting (reminds me of the Rollup refactor :-) ). That's why I hereby propose dumping complete backwards compatibility and bumping the Leaflet major version to 2. (ping @mourner @perliedman )

**Comment by johnd0e** [comment]
> That's why I hereby propose dumping complete backwards compatibility

I agree that this should be done, at some point.

But I still believe that we can solve most of the issues still keeping compatibility, that's the purpose of https://github.com/Leaflet/Leaflet/projects/1

**Comment by scaddenp** [comment]
Noting a problem from 7243:
Ok. I have found the source of the problem.
leaflet 1.4
var pointer = !!(window.PointerEvent || msPointer);
leaflet 1.6
var pointer = !webkit && !!(window.PointerEvent || msPointer);
webkit is true so pointer is false.
So when leaflet-draw try to register a listener for touchstart, (addOne), then pointer is false and so in
if (pointer && type.indexOf('touch') === 0) { // Needs DomEvent.Pointer.js addPointerListener(obj, type, handler, id);
addPointerListener is not called. No pointer events. Why do webkit browsers not receive pointer events?

**Comment by IvanSanchez** [comment]
@scaddenp That got removed in #7010, specifically https://github.com/Leaflet/Leaflet/pull/7010/files#diff-bc4c02f919ba9f4e3259e37fa12927a7R96

@johnd0e @mourner Maybe we should rush a 1.7 release to get #7010 out?

**Comment by scaddenp** [comment]
Looks like I need to check out master at least and check it runs with all my stuff. I have pretty big test suite.

**Comment by johnd0e** [comment]
> Maybe we should rush a 1.7 release to get #7010 out?

Sure, but someone should [prepare release](https://github.com/Leaflet/Leaflet/blob/master/RELEASE.md)..

**Comment by johnd0e** [comment]
@IvanSanchez 
Another case where touch events are still preferred: #7394

**Comment by jonkoops** [comment]
@IvanSanchez I think it's time we get to work on this. I'd like to get rid of `touchNative`, `touch` and `pointer` in the `Browser` namespace and I think it depends on this work.

What can we do to get this effort started?

**Comment by IvanSanchez** [comment]
> @IvanSanchez I think it's time we get to work on this. I'd like to get rid of `touchNative`, `touch` and `pointer` in the `Browser` namespace and I think it depends on this work.
> 
> What can we do to get this effort started?

Depends on how hard we want to go with the 2.0 breaking changes. Since we're deprecating browsers and we can assum `PointerEvent`s support, that simplifies things a lot.

Start by having a look at the [`pointerify` branch](https://github.com/Leaflet/Leaflet/commit/3151aadb17a64131ecaa533a1f328c6a41339ff1). In a nutshell:

- [ ] Replace all `mouse*` event names with `pointer*` 
  - [ ] Add a `ev.pointerType === 'mouse'` check, if appropriate
- [ ] Replace all `touch*` event names with `touch*`
  - [ ] Add a `ev.pointerType === 'touch'` check, if appropriate
- [ ] Optional: decorate event handlers received by `DomEvent.on`/`.off` (that was the goal of the "`unpointerify`" code in my branch)
  - [ ] Decorate handlers for `mouse*` handlers, e.g so that running something like `marker.on('mouseover', handler)` would actually behave like `marker.on('pointerover', (ev)=>ev.pointerType==='mouse' && handler(ev))`
  - [ ] Idem for `touch*` handlers

There is no need for decorating mouse/touch events as I was doing in the branch - the goal of that code was to support `PointerEvent`-like behaviour in browsers without native `PointerEvent`s.

**Comment by jonkoops** [comment]
Great, I'll see what I can do to get some work done for this. @Falke-Design do you perhaps feel like teaming up on this one? I think we could divvy up the tasks to replace existing `mouse*` events pretty easily and deliver incrementally.

**Comment by jonkoops** [comment]
We should probably also rename and refactor the following APIs to be more Pointer Event friendly:

- [`mouseEventToContainerPoint()`](https://leafletjs.com/reference.html#map-mouseeventtocontainerpoint)
- [`mouseEventToLayerPoint()`](https://leafletjs.com/reference.html#map-mouseeventtolayerpoint)
- [`mouseEventToLatLng()`](https://leafletjs.com/reference.html#map-mouseeventtolatlng)

@IvanSanchez WDYT?

**Comment by Falke-Design** [comment]
> Great, I'll see what I can do to get some work done for this. @Falke-Design do you perhaps feel like teaming up on this one? I think we could divvy up the tasks to replace existing `mouse*` events pretty easily and deliver incrementally.

Sounds good how do you want to work on this?



> We should probably also rename and refactor the following APIs to be more Pointer Event friendly:
> 
>     * [`mouseEventToContainerPoint()`](https://leafletjs.com/reference.html#map-mouseeventtocontainerpoint)
> 
>     * [`mouseEventToLayerPoint()`](https://leafletjs.com/reference.html#map-mouseeventtolayerpoint)
> 
>     * [`mouseEventToLatLng()`](https://leafletjs.com/reference.html#map-mouseeventtolatlng)

I think renaming makes sense but also a ton of project are using the functions and everyone needs to refactor this. Maybe we can support this as a wrapper? But on the other hand it would be a clean break with 2.0

**Comment by IvanSanchez** [comment]
On one hand it makes a lot of sense.

**But**, on the other hand, `PointerEvent` inherits from `MouseEvent`, and those functions still work with `MouseEvent`s. I'm unsure. :confused:

**Comment by Falke-Design** [comment]
My attempt to add touch events, maybe it helps for something: #7842

And some other event PRs: #7044, #7064

**Comment by jonkoops** [comment]
> Sounds good how do you want to work on this?

I think the best way is that we use this issue to communicate which files we're working on so we don't get in each other's way. I'd say we can pick something to work on and keep delivering until there are no references to mouse events left.

**Comment by Falke-Design** [comment]
- [ ] Control.Layers
- [ ] Events.leafdoc
- [ ] Layer.Interactive.leafdoc
- [ ] DomEvent
- [ ] DomEvent.DoubleTap
- [ ] DomEvent.Pointer
- [ ] Draggable
- [ ] ImageOverlay, VideoOverlay, SVGOverlay
- [ ] Layer: `bubblingMouseEvents`
- [ ] Tooltip
- [ ] Marker
- [ ] GridLayer
- [ ] Canvas
- [ ] Path: `bubblingMouseEvents`
- [ ] Map
- [x] Map.BoxZoom #8757
- [x] Map.Keyboard #8758 
- [ ] Map.ScrollWheelZoom
- [ ] Map.TabHold
- [ ] Map.TouchZoom

**Comment by Falke-Design** [comment]
Should we merge all this PRs into a seperate branch and at the end all together into `main`?

**Comment by jonkoops** [comment]
> Should we merge all this PRs into a seperate branch and at the end all together into `main`?

No, I think we can merge them straight into `main`. We're not planning a stable release of 2.0 anytime soon either way.

**Comment by joelhickok** [comment]
I guess the warning I see in Firefox from Util.js ties in directly to this:
>MouseEvent.mozPressure is deprecated. Use PointerEvent.pressure instead.

**Comment by syonfox** [comment]
Hey just going to work around it but regarding legacy plugin code. that im updating to support touch, I wanted to do this through pointer events since it is nice once we go through the growing pains. Its anoying to have more then one interface maybe. its more annoying to change :) anyway.

make sure that you still add the 

`containerPoint` to the event and also it looks like `map.on()` does not subscribe to "pointermove"


ie
```
map.on("pointermove", (e)=>{console.log(e)}) ; // kaka

document.addEventListener("pointermove", (e)=>{console.log(e)}) ; // works

```

<details> 
<summary> Use case </summary>

```js
       this.EVENT = {
            DOWN: "mousedown",
            UP: "mouseup",
            MOVE: "mousemove"
        }

        if(options.usePointer) {
             this.EVENT = {
            DOWN: "pointerdown",
            UP: "pointerup",
            MOVE: "pointermove"
        }
        }
        
        ...
        
        
 this._moved = false;
            let self = this
            this._downListener = (e) => {
                console.log(this.EVENT.DOWN)
                self._moved = false
                self._mousedown = true;

                self._map.on(this.EVENT.MOVE, self._moveListener)
                self._map.once(this.EVENT.UP, self._upListener)

                self.fire('resizeStart')
            }
            resizeHandle.addEventListener(this.EVENT.DOWN, this._downListener)
            this._moveListener = (e) => {
                console.log('move')
                console.log(e);

                if (self._mousedown) {

                    let height;
                    switch (this.options.position) {
                        case "top": {
                            height = e.containerPoint.y;
                            break;
                        }
                        case "bottom": {
                            height = self._map._container.clientHeight - e.containerPoint.y;
                            break;
                        }
                        case "left": {
                            height = e.containerPoint.x;
                            break;
                        }
                        case "right": {
                            height = self._map._container.clientWidth - e.containerPoint.x;
                            break;
                        }

                    }
                    self.resize(height);
                }
                self._moved = true
                // self._map.once(this.EVENT.UP, self._upListener)
            }
            // resizeHandle.addEventListener(this.EVENT.MOVE, this._moveListener)
            this._upListener = (e) => {
                self._mousedown = false;
                console.log("up")
                if (self._moved) {
                    console.log('moved')
                    console.log(e);
                } else {
                    console.log('not moved')
                }

                //this forces content to update
                // if(self.options.windowResize) {
                //     window.dispatchEvent(new Event('resize'));
                // }
                self._map.off(this.EVENT.MOVE, self._moveListener)
                self.fire('resizeEnd');
            }

        

```

</details>


... anyway I have no idea why the map.on("pointermove") is not firing the subscribed function probably a bug. 

my workaround for now is 
```js 
map._container.addEventListener('pointermove', e=>{
   e.containerPoint = map.mouseEventToContainerPoint(e); 
   console.log(e.containerPoint)
})
```

**Comment by Falke-Design** [comment]
@syonfox we are currently in the process to change all the events to `PointerEvents` but for now `pointermove` (and the same for all the other `pointer...` events) are not fully supported. Use instead `map.on("mousemove touchmove", (e)=>{console.log(e)}) ;`

**Comment by jonkoops** [comment]
Sorry that I have been afk from this discussion for a while, I've been very busy with my day job so I have not had the time to give this the attention I want.

@Falke-Design I am going to see if we can move your work forward, I'd like to work towards closing this issue so we can focus on landing 2.0.

**Comment by Falke-Design** [comment]
Sounds good! I will continue with it, if you are now able to review it 👍

**Comment by Falke-Design** [comment]
@jonkoops @IvanSanchez any suggestions how we should replace `e.touches`? With PointerEvents we don't have a list of touches anymore. Each pointer will fire a own `pointer`-event with a different `pointerId`. I think we need to create a list of pointers by our own. 

https://github.com/Leaflet/Leaflet/blob/51e380e45655ef44817dd8400c4e132c119a4775/src/map/handler/Map.TouchZoom.js#L41

**Comment by jonkoops** [comment]
@Falke-Design yeah, I cannot find a replacement for this either. I think we will have to keep track of all the active pointers on the Map ourselves.

**Comment by jonkoops** [comment]
This also seems to be the [recommended approach](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events/Pinch_zoom_gestures) on MDN.
